src/inmanta/protocol/ipc_light.py

```diff
@@ -0,0 +1,218 @@
+"""
+    Copyright 2024 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import abc
+import asyncio
+import logging
+import pickle
+import struct
+import uuid
+from asyncio import Future, Protocol, transports
+from dataclasses import dataclass
+from typing import Any, Callable, Coroutine, Optional
+
+
+class IPCException(Exception):
+    pass
+
+
+class ConnectionLost(IPCException):
+    pass
+
+
+@dataclass
+class IPCRequestFrame:
+    id: uuid.UUID
+    method: str
+    arguments: list[object]
```

Comment:
I wonder if we might improve on this by, instead of sending a method name and a list of arguments, to send a proper method request object, e.g.

```python
def f(x: int, y: str) -> None:
    pass

@dataclass
class IPCMethodCall: pass

@dataclass
class IPCRequestFrame:
    id: uuid.UUID
    method: IPCMethodCall

@dataclass
class IPCMethodCallF(IPCMethodCall):
    x: int
    y: str


def process_incoming_ipc_frame(frame: IPCRequestFrame) -> None:
    match frame.method:
        case IPCMethodCallF(x, y): f(x, y)
        case _: raise Exception()
```

> I wonder if we might improve on this

It's an open question. I haven't given it sufficient thought to be certain what benefit we get out of it. But I get the impression that it would give us a very low-effort typed interface.

Reply:
Even more interesting
```python
def f(x: int, y: str) -> None:
    pass


class IPCMethodCall(abc.ABC): 

     @abc.abstractmethod
     def call(self):
          ...

@dataclass
class IPCRequestFrame:
    id: uuid.UUID
    method: IPCMethodCall

@dataclass
class IPCMethodCallF(IPCMethodCall):
    x: int
    y: str
    
    def call():
        f(x,y)


def process_incoming_ipc_frame(frame: IPCRequestFrame) -> None:
   frame.method.call()
   
```

The big different is that the caller get's full control over what the server offers: if it can load it, it can execute it. 
Which is cool, but may be considered overpowered.

Reply:
We will keep this in mind for a next iteration, when we integrate